### ISSUE TITLE
TS should not allow assigning value with undefined union to non-nullable property when strictOptionalProperties

### ISSUE THREAD

**Issue by HerringtonDarkholme** [microsoft/TypeScript#44437]
# Bug Report
First, congrats to the new compiler switch! 
I don't know if this is intentional since strict optional property semantics seems tricky. Related to #44421.

### 🔎 Search Terms
strict optional properties, assignability

### 🕗 Version & Regression Information
Nightly, with strictOptionalProperties on

- I was unable to test this on prior versions because strictOptionalProperties is a new option

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?strict=true&strictOptionalProperties=true&ts=4.4.0-dev.20210602&ssl=10&ssc=22&pln=1&pc=1#code/CYUwxgNghgTiAEA3W8oC54G8DaAPDAzgC4wCWAdgOYC6G5ArgLYBGIM8APvMWVfAL4AoUJFgJk7Zhkzp4DFmwA08KdxIVKQkdDhIUYabPmsYy5gH5C6qlvA7xKYIbpMTZq70qd49cqABmFCDAAsJ2YnrsIM5yrkoqlmqe3r4BQSFCglDwALwq8AD0BfAA9gDWWbnwYIXF5fCVeSFF8GwwJTCNrbWtuAAO4EStMO0wQA)

### 💻 Code

```ts
declare var a: {[x: string]: number | string }
declare var b: {a: number, b: string}
declare var c: {a: number, b?: string}
declare var d: {a: number, b: string | undefined }
declare var e: {a: number, b?: string | undefined }

a = b // ok
a = c // ok 
a = d // error
a = e // expect error
```

### 🙁 Actual behavior

TS allow `a = e` where the property `b` might be undefined.

### 🙂 Expected behavior

TS should report error because of consistency?

**Comment by RyanCavanaugh** [comment]
@ahejlsberg I'm inclined to agree but let's discuss today
